proof/crefine/AARCH64/Syscall_C.thy

```diff
@@ -2229,26 +2229,42 @@ lemma ccorres_handleReservedIRQ:
                   simp flip: word_unat.Rep_inject split: if_splits)
   done
 
+(* Because of a case distinction, this proof occurs multiple times in handleInterrupt_ccorres below.
+   The proof is carefully manual to match up the C expression "if (!defined(CONFIG_ARM_GIC_V3))",
+   which renders as IF False or IF True in Simpl, with config_ARM_GIC_V3 from the spec side. *)
+method maybe_maskInterrupt_before_ack =
+    simp only: maskIrqSignal_def when_def,
+    rule ccorres_cond_seq,
+    rule ccorres_cond_both[where P="\<lambda>_. \<not>config_ARM_GIC_V3" and R=\<top>],
+       simp add: Kernel_Config.config_ARM_GIC_V3_def, (* match up !defined(..) condition *)
+      rule ccorres_gen_asm[where P="\<not>config_ARM_GIC_V3"],
+      simp,
+      ctac (no_vcg) add: maskInterrupt_ccorres,
+       ctac add: ackInterrupt_ccorres,
+      wp,
+     rule ccorres_gen_asm[where P="config_ARM_GIC_V3"],
+     simp,
+     ctac add: ackInterrupt_ccorres
+
 lemma handleInterrupt_ccorres:
   "ccorres dc xfdc
      (invs' and (\<lambda>s. irq \<in> non_kernel_IRQs \<longrightarrow> sch_act_not (ksCurThread s) s))
      \<lbrace>\<acute>irq = ucast irq\<rbrace>
      hs
      (handleInterrupt irq)
      (Call handleInterrupt_'proc)"
-  apply (cinit lift: irq_' cong: call_ignore_cong)
+  apply (cinit (no_ccorres_rewrite) lift: irq_' cong: call_ignore_cong)
```

Comment:
It might be worth a bit of a comment here about the `no_ccorres_rewrite`, like there is in `Interrupt_C.thy`. I don't think we've used it in our proofs before.

Reply:
Fair, there should be the same kind of comment here, yes. Raf had an idea for maybe wrapping the ‘defined’ such that it is not being rewritten, but I’d like to try that after we merge.

Reply:
Yes, please see my writeup in the comment in Interrupt_C ... I'm not a fan of disabling ccorres_rewrite like this, and believe there are other ways towards stability on the C side.